### PR TITLE
Fix Issue 15103 - Improve declaration / initialization syntax error message

### PR DESCRIPTION
**Pull request by adelavais**

### CODE REVIEW

File: src/dmd/parse.d [hunk 1]
```diff
@@ -6743,6 +6751,12 @@ final class Parser(AST) : Lexer
         return parseAssignExp();
     }
 
+    private void check()
```
Comment (MoonlightSentinel) [inline src/dmd/parse.d:6754]: Simply inline this function, it's not used anywhere else

File: src/dmd/parse.d [hunk 2]
```diff
@@ -6743,6 +6751,12 @@ final class Parser(AST) : Lexer
...
+    {
+        error("incorrect instantiation of `%s`", token.toChars());
```
Comment (MoonlightSentinel) [inline src/dmd/parse.d:6756]: This error message is not really helpful, maybe smth like this is more appropriate:

Given `Foo foo(42);`
```
foo.d(1): Found `(` when expecting `;` or `=`, did you mean `Foo foo = 42`?
```

**Review by MoonlightSentinel — DISMISSED** [review]
Thanks for your PR.

File: src/dmd/parse.d (hunk 2 above)
Comment (adelavais) [reply src/dmd/parse.d]: Thank you for your feedback, I made the modifications!

File: src/dmd/parse.d [hunk 3]
```diff
@@ -5533,7 +5533,18 @@ final class Parser(AST) : Lexer
...
                 AST.Expression exp = parseExpression();
-                check(TOK.semicolon, "statement");
+                /* https://issues.dlang.org/show_bug.cgi?id=15103
+                 * Improve declaration / initialization syntax error message
+                 * Error: found 'foo' when expecting ';' following statement
+                 * becomes Error: Found `(` when expecting `;` or `=`, did you mean `Foo foo = 42`?
+                 */
+                if (token.value == TOK.identifier)
+                {
+                    error("Found `%s` when expecting `;` or `=`, did you mean `%s %s = %s`?", peek(&token).toChars(), exp.toChars(), token.toChars(), peek(peek(&token)).toChars());
```
Comment (NilsLankila) [inline src/dmd/parse.d:5543]: `Found` should not start with a capitalized char.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Please squash the commits and address Nils's comment and we are good to go.

File: test/fail_compilation/fail196.d [hunk 4]
```diff
@@ -10,10 +10,10 @@ fail_compilation/fail196.d(29): Error: found `";\x0a    assert(s == "` when expe
 fail_compilation/fail196.d(30): Error: found `");\x0a\x0a    s = q"` when expecting `;` following statement
 fail_compilation/fail196.d(32): Error: found `{` when expecting `;` following statement
 fail_compilation/fail196.d(32): Error: found `}` when expecting `;` following statement
-fail_compilation/fail196.d(33): Error: found `foo` when expecting `;` following statement
+fail_compilation/fail196.d(33): Error: Found `{` when expecting `;` or `=`, did you mean `";\x0a    assert(s == " foo = xxx`?
 fail_compilation/fail196.d(33): Error: found `}` when expecting `;` following statement
 fail_compilation/fail196.d(35): Error: found `<` when expecting `;` following statement
-fail_compilation/fail196.d(36): Error: found `foo` when expecting `;` following statement
+fail_compilation/fail196.d(36): Error: Found `<` when expecting `;` or `=`, did you mean `xxx >> ";\x0a    assert(s == " foo = xxx`?
```
Comment (ibuclaw) [inline test/fail_compilation/fail196.d:16]: I know the code in this test is nonsense anyway, but perhaps this is a small sign that the suggestion style message should be more restricted to parsed tokens that could make sense in the rewritten form (a shift expression can never be a type).

**Comment by adelavais** [comment]
> Please squash the commits and address Nils's comment and we are good to go.

The above is my attempt to squash. Sorry about that. Will redo.

File: src/dmd/parse.d [hunk 5]
```diff
@@ -5533,7 +5533,18 @@ final class Parser(AST) : Lexer
         Lexp:
             {
                 AST.Expression exp = parseExpression();
-                check(TOK.semicolon, "statement");
+                /* https://issues.dlang.org/show_bug.cgi?id=15103
+                 * Improve declaration / initialization syntax error message
+                 * Error: found 'foo' when expecting ';' following statement
+                 * becomes Error: found `(` when expecting `;` or `=`, did you mean `Foo foo = 42`?
+                 */
+                if (token.value == TOK.identifier)
```
Comment (ibuclaw) [inline src/dmd/parse.d:5541]:
```suggestion
                if (token.value == TOK.identifier && exp.op == TOK.identifier)
```

Comment (ibuclaw) [reply src/dmd/parse.d:5541]: Then revert the message change in fail196.d

**Review by Geod24 — APPROVED** [review]

File: src/dmd/parse.d (hunk 5 above)
Comment (adelavais) [reply src/dmd/parse.d]: Thank you for your feedback! I modified.

**Review by MoonlightSentinel — APPROVED** [review]

**Review by ibuclaw — APPROVED** [review]

**Pull request merged**